cmd/state-svc/internal/resolver/analytics/analytics.go

```diff
@@ -219,12 +227,52 @@ func (r *Resolver) flush() error {
 		return errs.Wrap(err, "Failed to load deferred events")
 	}
 	for _, event := range events {
-		r.event(event)
+		ev := &event
+		r.addProjectIDToEvent(ev)
+		r.event(ev)
 	}
 
 	return nil
 }
 
+func (r *Resolver) addProjectIDToEvent(ev *event.EventData) {
+	if ev.ProjectName == "" || ev.ProjectID != "" {
+		return
+	}
+
+	ev.ProjectID = r.projectID(ev.ProjectName)
+}
+
+// projectID resolves the projectID from projectName and caches the result in the provided projectIDMap
+func (r *Resolver) projectID(projectName string) string {
```

Comment:
Nit: `projectName` is the namespace I assume? Would prefer we name this something that conveys that (goes for other places where this is used also).